### PR TITLE
Introduce Async/Await for rust-spdm.

### PR DESCRIPTION
**Pull request by longlongyang**

### CODE REVIEW

File: executor/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,38 @@
+// Copyright (c) 2023 Intel Corporation
+//
+// SPDX-License-Identifier: BSD-2-Clause-Patent
```
Comment (jyao1) [inline executor/src/lib.rs]: please use apache 2.0 license

Comment (longlongyang) [reply executor/src/lib.rs]: > please use apache 2.0 license

Updated.

File: sh_script/build.sh [hunk 2]
```diff
@@ -2,6 +2,9 @@
 
 set -euo pipefail
 
+export RUST_MIN_STACK=10485760
+set RUST_MIN_STACK=10485760
```
Comment (jyao1) [inline sh_script/build.sh:6]: how this is calculated?

File: test/spdm-requester-emu/src/main.rs [hunk 3]
```diff
@@ -26,54 +27,78 @@ use spdm_emu::socket_io_transport::SocketIoTransport;
...
 
-fn send_receive_hello(
-    stream: &mut TcpStream,
-    transport_encap: &mut dyn common::SpdmTransportEncap,
+use spin::Mutex;
+extern crate alloc;
+use alloc::sync::Arc;
+use core::ops::DerefMut;
+
+const EMU_STACK_SIZE: usize = 10;
```
Comment (jyao1) [inline test/spdm-requester-emu/src/main.rs]: how this is calculated?

File: test/spdm-requester-emu/src/main.rs [hunk 4]
```diff
@@ -369,10 +418,30 @@ fn main() {
...
-    test_spdm(socket_io_transport, transport_encap);
+    executor::block_on(send_receive_stop(socket, transport_encap, transport_type));
+}
 
-    send_receive_stop(&mut socket, transport_encap, transport_type);
+fn main() {
+    use std::thread;
+
+    thread::Builder::new()
+        .stack_size(EMU_STACK_SIZE * 1024 * 1024)
```
Comment (jyao1) [inline test/spdm-requester-emu/src/main.rs]: This is confusing.
Why not define EMU_STACK_SIZE to be `10 * 1024 * 1024` directly?

File: sh_script/build.sh (hunk 2 above)
Comment (longlongyang) [reply sh_script/build.sh:6]: > how this is calculated?

This is for Fuzz code to pass build and run, just hardcode it in script.

File: test/spdm-requester-emu/src/main.rs (hunk 3 above)
Comment (longlongyang) [reply test/spdm-requester-emu/src/main.rs]: > how this is calculated?

updated.

File: test/spdm-requester-emu/src/main.rs (hunk 4 above)
Comment (longlongyang) [reply test/spdm-requester-emu/src/main.rs]: > This is confusing. Why not define EMU_STACK_SIZE to be `10 * 1024 * 1024` directly?

Good advice, updated.

File: test/spdm-responder-emu/src/main.rs [hunk 5]
```diff
@@ -25,10 +36,104 @@ use spdm_emu::socket_io_transport::SocketIoTransport;
...
+use core::ops::DerefMut;
+
+// the possible path to create max stack consumption is either handle_spdm_certificate or handle_spdm_measurement in session.
+// there is no harm to sum these two to be the target stack size preservation.
+const EMU_STACK_SIZE: usize = size_of::<spdmlib::common::SpdmContext>() +
+                            size_of::<SpdmCertChainData>() * (SPDM_MAX_SLOT_NUMBER + 1) + size_of::<SpdmCertChainBuffer>() * SPDM_MAX_SLOT_NUMBER + // SpdmProvisionInfo
+                            size_of::<SpdmCertChainData>() * (SPDM_MAX_SLOT_NUMBER + 1) + // SpdmPeerInfo
+                            size_of::<SocketIoTransport>() + // SpdmDeviceIo
+                            size_of::<PciDoeTransportEncap>() + // PciDoeTransportEncap
+                            size_of::<MctpTransportEncap>() + // SpdmTransportEncap
```
Comment (jyao1) [inline test/spdm-responder-emu/src/main.rs]: please use MAX(PciDoeTransportEncap, MctpTransportEncap).
Also, we can define a new macro for transport required size.

File: test/spdm-responder-emu/src/main.rs [hunk 6]
```diff
@@ -25,10 +36,104 @@ use spdm_emu::socket_io_transport::SocketIoTransport;
...
+                            MAX_SPDM_MEASUREMENT_RECORD_SIZE * 255 + // ~ measurement_collection
+                            size_of::<ManagedBuffer12Sign>() + // generate_measurement_signature
+                            size_of::<SpdmSignatureStruct>() + // generate_measurement_signature
+                            SENDER_BUFFER_SIZE + // send_secured_message
+                            size_of::<usize>() * 2 + // send_secured_message
+                            SENDER_BUFFER_SIZE + // encode_secured_message
+                            SENDER_BUFFER_SIZE + // encode_secured_message
+                            size_of::<usize>() * 7 + // encode_secured_message
+                            SENDER_BUFFER_SIZE + // encode_msg
+                            size_of::<usize>() * 13 + // encode_msg
```
Comment (jyao1) [inline test/spdm-responder-emu/src/main.rs]: there are too many details. Please simplify that.

File: test/spdm-responder-emu/src/main.rs [hunk 7]
```diff
@@ -25,10 +36,104 @@ use spdm_emu::socket_io_transport::SocketIoTransport;
...
+                            RECEIVER_BUFFER_SIZE + // decode_secured_message
+                            size_of::<usize>() * 7 + // decode_secured_message
+                            RECEIVER_BUFFER_SIZE + // decode_msg
+                            size_of::<usize>() * 10 + // decode_msg
+                            size_of::<Reader>() * 2 + // decode_msg
+                            size_of::<SpdmDigestStruct>() + // decode_msg
+                            size_of::<usize>() * 4 + // receive_message in spdm_emu
+                            size_of::<Reader>() + // receive_message in spdm_emu
+                            size_of::<SpdmSocketHeader>() + // receive_message in spdm_emu
+                            2 * 1024 * 1024; // for emu stack
```
Comment (jyao1) [inline test/spdm-responder-emu/src/main.rs]: what content is this 2M?
Also, please define a new macro for that.

File: test/spdm-responder-emu/src/main.rs (hunk 5 above)
Comment (longlongyang) [reply test/spdm-responder-emu/src/main.rs]: > please use MAX(PciDoeTransportEncap, MctpTransportEncap). Also, we can define a new macro for transport required size.

updated.

File: test/spdm-responder-emu/src/main.rs (hunk 6 above)
Comment (longlongyang) [reply test/spdm-responder-emu/src/main.rs]: > there are too many details. Please simplify that.

simplified on latest commit.

File: test/spdm-responder-emu/src/main.rs (hunk 7 above)
Comment (longlongyang) [reply test/spdm-responder-emu/src/main.rs]: > what content is this 2M? Also, please define a new macro for that.

Removed the estimated size, in latest commit, there is a formula to calculate it.

**Pull request merged**